src/puya/ir/optimize/compiled_reference.py

```diff
@@ -0,0 +1,163 @@
+        self,
+        const: ir.CompiledContractReference,
+    ) -> ir.CompiledContractReference | ir.Constant:
+        field = const.field
+        if field in (
+            TxnField.GlobalNumUint,
+            TxnField.GlobalNumByteSlice,
+            TxnField.LocalNumUint,
+            TxnField.LocalNumByteSlice,
+        ):
+            try:
+                state_total = self.context.state_totals[const.artifact]
+            except KeyError:
+                raise InternalError(
+                    f"Invalid contract reference: {const.artifact}", const.source_location
+                ) from None
+            match field:
+                case TxnField.GlobalNumUint:
+                    total = state_total.global_uints
+                case TxnField.GlobalNumByteSlice:
+                    total = state_total.global_bytes
+                case TxnField.LocalNumUint:
+                    total = state_total.local_uints
+                case TxnField.LocalNumByteSlice:
+                    total = state_total.local_bytes
+                case _:
+                    raise InternalError(
+                        f"Invalid state total field: {field.name}", const.source_location
+                    )
+            return ir.UInt64Constant(
+                value=total,
+                source_location=const.source_location,
+            )
+
+        if not _is_constant(const.template_variables):
+            return const
+        template_constants = _get_template_constants(
+            self.context.options.template_variables, const.template_variables
+        )
+        match field:
+            case TxnField.ApprovalProgramPages | TxnField.ClearStateProgramPages:
+                page = const.program_page
+                if page is None:
+                    raise InternalError("expected non-none value for page", const.source_location)
+                program_bytecode = self.context.get_program_bytecode(
+                    const.artifact,
+                    ("approval" if field == TxnField.ApprovalProgramPages else "clear_state"),
+                    template_constants,
+                )
+                program_page = program_bytecode[
+                    page * MAX_BYTES_LENGTH : (page + 1) * MAX_BYTES_LENGTH
+                ]
+                return ir.BytesConstant(
+                    value=program_page,
+                    encoding=AVMBytesEncoding.base64,
+                    source_location=const.source_location,
+                )
+            case TxnField.ExtraProgramPages:
+                approval_bytecode = self.context.get_program_bytecode(
+                    const.artifact, "approval", template_constants
+                )
+                clear_bytecode = self.context.get_program_bytecode(
+                    const.artifact, "clear_state", template_constants
+                )
+                total_bytes = len(approval_bytecode) + len(clear_bytecode)
```

Comment:
surprised we don't already have a utility method for this calculation - maybe this should become one?

Reply:
Moved it into a util function